app/views/unattended/provisioning_templates/snippet/saltstack_minion.erb

```diff
@@ -20,4 +23,12 @@ log_level: warning
 # * {'roles': ['webserver', 'frontend']}
 #
 %>
-grains: <%= host_param('salt_grains') ? host_param('salt_grains') : '{}' %>
+<% collected_grains = @host.derive_salt_grains(use_autosign: true) %>
+<% if collected_grains.empty? -%>
+grains: {}
+<% else -%>
+grains: <% collected_grains.each do |key, val| %>
+  <%= "'#{key}'" -%>: <%= "'#{val}'" -%>
+  <% end -%>
+<% end -%>
```

Comment:
I guess, because the keys/values need to be quoted.

Reply:
I don't like guessing?
What if the value contains a `'`?
`.to_json` or `.to_yaml` should quote stuff properly.

Reply:
good your point. Thanks @evgeni 

Would you be ok with the following changes:
1. Template change. Use this one:
```
grains:  
<%= indent(2) { @host.derive_salt_grains_yaml(use_autosign: true) } %>
```
2. Add the method derive_salt_grains_yaml in foreman_salt which returns yaml?

(to_yaml is not possible because of safe mode rendering)

Reply:
how about `to_yaml(@host…)`

I thought that'd be allowed in safe mode (but it is also calling `obj.to_yaml`)?

https://github.com/theforeman/foreman/blob/35076bff03e78ab26742b3b133d231ca29445101/lib/foreman/renderer/scope/macros/helpers.rb#L68-L75

Reply:
This is perfect. So the code would look like:

grains:  
<%= indent(2) { to_yaml(@host.derive_salt_grains(use_autosign: true)).gsub(/---\n/, "") } %>

(and works in safemode rendering)

Reply:
Just tried, `to_yaml({'lol':'what'})` works just fine, while `{'lol':'what'}.to_yaml` raises `undefined method '#to_yaml' for Hash::Jail (Hash)`

Reply:
I would actually do:

```erb
grains: <%= to_json(@host.derive_salt_grains(use_autosign: true)) %>
```

As any valid JSON is also valid YAML and you save this ugly `gsub` and `indent`?

Reply:
And if you wonder about the rendering:

empty hash:
```yaml
grains: {
}
```

non empty:
```yaml
grains: {
  "lol": "what",
  "dit": "dat"
}
```

Reply:
thank you. very appreciated!